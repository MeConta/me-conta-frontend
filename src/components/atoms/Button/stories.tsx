import { Story, Meta } from '@storybook/react'

import { Button, ButtonProps } from '.'

export default {
  component: Button,
  title: 'Atoms/Button',
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />
export const asLink: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  color: 'primary',
  size: 'large',
  radius: 'round',
  children: 'Button'
}

asLink.args = {
  color: 'primary',
  size: 'large',
  radius: 'round',
  children: 'Button',
  as: 'a',
  href: '/link'
}
