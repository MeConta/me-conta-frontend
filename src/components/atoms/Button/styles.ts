import styled, { css, DefaultTheme } from 'styled-components'
import Animation from '../../../utils/animations/animation'
import { ButtonProps } from '.'

type WrapperProps = Pick<
  ButtonProps,
  | 'size'
  | 'color'
  | 'radius'
  | 'textTransform'
  | 'btnStyle'
  | 'isLoading'
  | 'fillOver'
  | 'fillOverDuration'
>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    min-height: 1.5rem;
    font-size: ${theme.font.sizes['desk-large']};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    min-height: 3rem;
    font-size: ${theme.font.sizes['desk-medium']};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    min-height: 5rem;
    font-size: ${theme.font.sizes['desk-xlarge']};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  mediumLarge: (theme: DefaultTheme) => css`
    min-height: 6rem;
    font-size: ${theme.font.sizes['desk-glarge']};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  primary: (theme: DefaultTheme) => css`
    background: ${theme.colors.cornflowerBlue};

    &:hover {
      background-color: ${theme.colors.cobalt};
      transition: background-color 0.2s;
    }
  `,
  secondary: (theme: DefaultTheme) => css`
    background: ${theme.colors.ceriseRed};

    &:hover {
      background-color: ${theme.colors.venetianRed};
      transition: background-color 0.2s;
    }
  `,
  negative: (theme: DefaultTheme) => css`
    background: ${theme.colors.white};
    color: ${theme.colors.ceriseRed};
    border: 1px solid ${theme.colors.ceriseRed};

    &:hover {
      color: ${theme.colors.maroonFlush};
      transition: color 0.2s;
    }
  `,
  round: (theme: DefaultTheme) => css`
    border-radius: ${theme.border['btn-round-radius']};
  `,
  square: (theme: DefaultTheme) => css`
    border-radius: ${theme.border['btn-square-radius']};
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
      transition: filter 0.2s;
    }
  `,
  uppercase: () => css`
    text-transform: uppercase;
  `,
  link: (theme: DefaultTheme) => css`
    background-color: transparent;
    color: ${theme.colors.venetianRed};
    font-size: 1.5rem;
    text-decoration: underline;
    &:hover {
      background-color: transparent;
    }
  `,
  fillOver: (duration: number) => css`
    ::before {
      content: '';
      background-color: rgba(0, 0, 0, 0.2);
      position: absolute;
      right: 100%;
      width: 100%;
      height: 100%;
      animation: slideIn ${duration}ms linear forwards;
      @keyframes slideIn {
        100% {
          right: 0;
        }
      }
    }
  `
}

const [fadeIn, fadeOut] = [
  Animation.absolute()
    .right({ initialPosition: '50px', finalPosition: '8px' })
    .setAnimation()
    .fadeIn(),
  Animation.absolute()
    .right({ initialPosition: '8px', finalPosition: '50px' })
    .setAnimation()
    .fadeOut()
]

export const Wrapper = styled.button<WrapperProps>`
  ${({
    theme,
    size,
    color,
    radius,
    disabled,
    textTransform,
    btnStyle,
    isLoading,
    fillOver,
    fillOverDuration
  }) => css`
    overflow-x: hidden;
    align-items: center;
    border: 0;
    color: ${theme.colors.white};
    display: inline-flex;
    justify-content: center;
    text-decoration: none;
    cursor: pointer;
    text-align: center;
    font-family: Mulish;
    font-weight: ${theme.font.bold};
    position: relative;
    > svg {
      width: ${theme.font.sizes['desk-large']};
      height: ${theme.font.sizes['desk-large']};
      margin-right: ${theme.spacings.xxxsmall};
    }
    > #loader {
      ${isLoading ? fadeIn : fadeOut}
    }

    ${!!size && wrapperModifiers[size](theme)};
    ${!!color && wrapperModifiers[color](theme)};
    ${!!radius && wrapperModifiers[radius](theme)};
    ${disabled && wrapperModifiers.disabled()};
    ${!!textTransform && wrapperModifiers[textTransform]};
    ${!!btnStyle && wrapperModifiers[btnStyle](theme)};
    ${fillOver && wrapperModifiers.fillOver(fillOverDuration!)};
  `}
`
export const TextWrapper = styled.div`
  position: relative;
`
