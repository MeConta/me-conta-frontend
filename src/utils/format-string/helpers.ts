// Rever tipagem
export const formatPhoneNumber = (value: string): any =>
  value
    .replace(/\D/g, '')
    .replace(/(\d{2})(\d)/, '($1) $2')
    .replace(/(\s\d{1})/, '$1 ')
    .replace(/(\d{4})(\d)/, '$1-$2')
    .replace(/(-\d{4})(\d+?)$/, '$1')

export const removePhoneMask = (value: string) => value.replace(/\D/g, '')
