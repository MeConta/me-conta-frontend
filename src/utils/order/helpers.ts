import { VolunteerAvailableSlot } from '../../services/volunteers-service/volunteer-service'

function convertSlotsIntoString(dateSlot: VolunteerAvailableSlot) {
  return new Date(dateSlot.inicio).toLocaleTimeString([], {
    hour: '2-digit',
    minute: '2-digit',
    hour12: false
  })
}

export const sortSlotsByTime = (dateSlot: VolunteerAvailableSlot[]) => {
  return dateSlot.sort((a, b) => {
    const x = convertSlotsIntoString(a)
    const y = convertSlotsIntoString(b)
    return x == y ? 0 : x > y ? 1 : -1
  })
}
