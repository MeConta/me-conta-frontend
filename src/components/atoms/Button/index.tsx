import { AnchorHTMLAttributes, ButtonHTMLAttributes, ElementType } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  children: React.ReactNode
  size?: 'medium' | 'large'
  color?: 'primary' | 'secondary' | 'negative'
  radius?: 'round' | 'square'
  as?: ElementType
} & ButtonTypes

export function Button({
  children,
  size = 'large',
  color = 'primary',
  radius = 'round',
  ...props
}: ButtonProps) {
  return (
    <S.Wrapper size={size} color={color} radius={radius} {...props}>
      {children}
    </S.Wrapper>
  )
}
