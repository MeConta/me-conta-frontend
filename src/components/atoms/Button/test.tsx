import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import Button from '.'

describe('<Button/>', () => {
  it('should render large size by default', () => {
    renderWithTheme(<Button>button</Button>)
    expect(screen.getByRole('button')).toHaveStyle({
      height: '5rem',
      'font-size': '2.2rem'
    })
    expect(screen.getByRole('button')).toBeInTheDocument()
    expect(screen.getByRole('button')).toMatchSnapshot()
  })

  it('should render button with color', () => {
    renderWithTheme(<Button>button</Button>)
    expect(screen.getByRole('button')).toHaveStyle({ background: '#458FF6' })
  })

  it('should render button with color when hovering over', () => {
    renderWithTheme(<Button>button</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule(
      'background-color',
      '#0545AB',
      { modifier: ':hover' }
    )
  })

  it('should render the button with a border radius', () => {
    renderWithTheme(<Button>button</Button>)
    expect(screen.getByRole('button')).toHaveStyle({
      'border-radius': '3.4rem'
    })
  })

  it('should render the button with a medium size', () => {
    renderWithTheme(<Button size="medium">button</Button>)
    expect(screen.getByRole('button')).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render a button with secondary color', () => {
    renderWithTheme(<Button color="secondary">button</Button>)
    expect(screen.getByRole('button')).toHaveStyle({
      'background-color': '#DE3163'
    })
  })
})
