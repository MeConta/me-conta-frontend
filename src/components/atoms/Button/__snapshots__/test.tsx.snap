// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render large size by default 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  text-align: center;
  font-family: Mulish;
  font-weight: 700;
  min-height: 5rem;
  font-size: 2.2rem;
  padding: 0.8rem 4.8rem;
  background: #458FF6;
  border-radius: 3.4rem;
}

.c0:hover {
  background-color: #0545AB;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
}

<button
  class="c0"
  color="primary"
  radius="round"
>
  button
</button>
`;
